server/events/project_command_builder.go

```diff
@@ -483,26 +485,50 @@ func (p *DefaultProjectCommandBuilder) getTfVersion(ctx *CommandContext, absProj
 		return nil
 	}
 
-	if len(module.RequiredCore) != 1 {
-		ctx.Log.Info("cannot determine which version to use from terraform configuration, detected %d possibilities.", len(module.RequiredCore))
+	if len(module.RequiredCore) == 0 {
+		ctx.Log.Debug("no version constraints detected from terrafrom configuration.")
 		return nil
 	}
-	requiredVersionSetting := module.RequiredCore[0]
 
-	// We allow `= x.y.z`, `=x.y.z` or `x.y.z` where `x`, `y` and `z` are integers.
-	re := regexp.MustCompile(`^=?\s*([^\s]+)\s*$`)
-	matched := re.FindStringSubmatch(requiredVersionSetting)
-	if len(matched) == 0 {
-		ctx.Log.Debug("did not specify exact version in terraform configuration, found %q", requiredVersionSetting)
+	// build constraints
+	constraintStrings := strings.Join(module.RequiredCore, ",")
+
+	constraints, err := semver.NewConstraint(constraintStrings)
+	if err != nil {
+		ctx.Log.Err("trying to build constraints: %s", err)
 		return nil
 	}
-	ctx.Log.Debug("found required_version setting of %q", requiredVersionSetting)
-	version, err := version.NewVersion(matched[1])
+
+	// build versions
+	releaseList, err := p.ReleasesLister.ListReleases()
 	if err != nil {
-		ctx.Log.Debug(err.Error())
+		ctx.Log.Err("trying to get list of terraform versions: %s", err)
 		return nil
 	}
 
-	ctx.Log.Info("detected module requires version: %q", version.String())
-	return version
+	versions := make([]*semver.Version, len(releaseList))
+	for i, release := range releaseList {
+		version, err := semver.NewVersion(release)
+		if err != nil {
+			ctx.Log.Err("trying to build version list: %s", err)
+			return nil
+		}
+
+		versions[i] = version
+	}
+
+	// sortVersions
+	sort.Sort(sort.Reverse(semver.Collection(versions)))
+
+	for _, version := range versions {
+		if constraints.Check(version) {
+			ctx.Log.Info("detected module requires version: %q", version)
+			return version
+		}
+
+	}
```

Comment:
Doesn't this change the implementation to basically allow more than just "=" version constraints?  I don't think we want to just silently upgrade/downgrade the terraform version for a workspace based on the constraints.  

Please tell me if im not understanding this correctly.

Reply:
@nishkrishnan This allows for us to use any version constraints supported by Terraform (which follows the SemVer spec). The expected behaviour is that I can use any of the SemVer operators allowed by Terraform and that Atlantis will use the correct version based on the SemVer constraint.

If a user wants to use an explicit version they can do that by using the `=` operator. If they want to use the most recent patch version of `0.12` for example they could specify `~> 0.12.0`. You can read more in [my comment](https://github.com/runatlantis/atlantis/issues/1217#issuecomment-726378380) on the issue.